### PR TITLE
fix: prevent overwrite of hidden relations in patch

### PR DESCRIPTION
**Pull request by Yelinz**
Thanks for @anehx pointing this out and @open-dynaMIX suggesting a solution path.

### CODE REVIEW

File: generic_permissions/visibilities.py [hunk 1]
```diff
@@ -60,6 +60,30 @@ def get_attribute(self, instance):
         return queryset
 
 
+class VisibilitySerializerMixin:
```
Comment (anehx) [inline generic_permissions/visibilities.py:63]: This would mean that all serializers using `VisibilityManyRelatedField` need to inherit from this mixin right? If so, we should add documentation for this.

Comment (Yelinz) [reply generic_permissions/visibilities.py:63]: Added a line in the README.

> To prevent that include the VisibilitySerializerMixin in your serializer where you use ManyToMany relations.

File: generic_permissions/visibilities.py [hunk 2]
```diff
@@ -60,6 +60,30 @@ def get_attribute(self, instance):
...
+                type(field) is not VisibilityManyRelatedField
+                or key not in validated_data
+            ):
+                continue
+
+            queryset = getattr(self.instance, key).all()
+            for handler in VisibilitiesConfig.get_handlers(queryset.model):
+                queryset = handler(queryset, self.context)
+
+            validated_data[key] += getattr(self.instance, key).exclude(pk__in=queryset)
```
[the remark this answers is not in the document]
Comment (Yelinz) [reply generic_permissions/visibilities.py]: Added some, check if its enough.

File: generic_permissions/visibilities.py (hunk 1 above)
Comment (anehx) [reply generic_permissions/visibilities.py:63]: :+1:

File: generic_permissions/visibilities.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anehx) [reply generic_permissions/visibilities.py]: :+1:

**Review by anehx — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -76,15 +73,21 @@ class MyModelViewset(VisibilityViewMixin, ModelViewSet):
 Data leaks in REST Framework may happen if you use includes (or even only references) from a model the user may see to something that should be hidden.
 To avoid such leaks, make sure to use a subclassed related field (either by creating your own using the provided `VisibilityRelatedFieldMixin`, or by using one of the provided types. See example below).
 
+Similar problems can occur when PATCHing existing relationships where some are hidden through the visibilities.
+The hidden relations will be overwritten by the latest update.
+To prevent that include the `VisibilitySerializerMixin` in your serializer where you use `ManyToMany` relations.
+Preferably after the `ValidatorMixin` so that validations first run and only afterwards it considers PATCHing the relation.
```
Comment (winged) [inline README.md:79]: I'd rephrase this a bit. What exactly happens? Are the hidden relations simply dropped? Also I'd formulate it in a way that "if you use the `VisibilityRelatedFieldMixin`, you will need to use the  `VisibilitySerializerMixin`..."

Also drop the "preferably" - I think this is a *must*

File: generic_permissions/visibilities.py [hunk 4]
```diff
@@ -60,6 +60,33 @@ def get_attribute(self, instance):
         return queryset
 
 
+class VisibilitySerializerMixin:
```
Comment (winged) [inline generic_permissions/visibilities.py:63]: Please add a docstring, I think it's important people understand this!

File: generic_permissions/visibilities.py [hunk 5]
```diff
@@ -60,6 +60,33 @@ def get_attribute(self, instance):
...
+            for handler in VisibilitiesConfig.get_handlers(queryset.model):
+                queryset = handler(queryset, self.context)
+
+            # Add remaining relations which can not be included in the request.
+            validated_data[key] += getattr(self.instance, key).exclude(pk__in=queryset)
+
+        return validated_data
+
+
 class VisibilityRelatedFieldMixin:
```
Comment (winged) [inline generic_permissions/visibilities.py:90]: I would recommend adding a check here to ensure the serializer has the (strongly recommended) `VisibilitySerializerMixin` in it's base class list:

```suggestion
class VisibilityRelatedFieldMixin:
    def bind(self, field_name, parent):
        parent_cls = type(parent)
        if VisibilitySerializerMixin not in parent_cls.mro():
            raise RuntimeWarning(
                f"To avoid data loss, consider using VisibilitySerializerMixin in {parent_cls.__name__}")
        return super().bind(field_name, parent)

```

**Review by winged — CHANGES_REQUESTED** [review]
Mostly some docs that I'd recommend adding, but also some validation to ensure people don't just accidentally use DGAP wrong and lose data

File: generic_permissions/visibilities.py (hunk 4 above)
Comment (Yelinz) [reply generic_permissions/visibilities.py:63]: Done

File: README.md (hunk 3 above)
Comment (Yelinz) [reply README.md]: Done

File: generic_permissions/visibilities.py (hunk 5 above)
Comment (Yelinz) [reply generic_permissions/visibilities.py:103]: Added

File: generic_permissions/visibilities.py [hunk 6]
```diff
@@ -60,6 +60,46 @@ def get_attribute(self, instance):
...
+        validated_data = super().validate(*args, **kwargs)
+
+        if not self.instance:
+            return validated_data
+
+        # Incoming patches can have a subset of all relations, so we need to
+        # keep the existing relations which are not included in the request.
+        for key, field in self.fields.items():
+            if (
+                type(field) is not VisibilityManyRelatedField
```
Comment (winged) [inline generic_permissions/visibilities.py:87]: You should use `isinstance` here, not a direct type check

```suggestion
                not isinstance(field, VisibilityManyRelatedField)
```

File: generic_permissions/visibilities.py [hunk 7]
```diff
@@ -88,6 +128,17 @@ def get_attribute(self, instance):
 
         return model_instance
 
+    def bind(self, field_name, parent):
+        parent_cls = type(parent)
+        if parent_cls is VisibilityManyRelatedField:
+            return super().bind(field_name, parent)
```
Comment (winged) [inline generic_permissions/visibilities.py:134]: Should probably use `isinstance()` here

File: generic_permissions/visibilities.py [hunk 8]
```diff
@@ -88,6 +128,17 @@ def get_attribute(self, instance):
...
+
+        if VisibilitySerializerMixin not in parent_cls.mro():
```
Comment (winged) [inline generic_permissions/visibilities.py:136]: Just thinking about it a bit more, you could use `isinstance()` here as well:

```suggestion
        if not isinstance(parent, VisibilitySerializerMixin):
```

It's basically the same: `isinstance()` just checks the method resolution order (MRO) of the associated class

**Review by winged — DISMISSED** [review]
Awesome work cleaning that up! Just some last things noticed in my suggested code

File: generic_permissions/visibilities.py (hunk 6 above)
Comment (Yelinz) [reply generic_permissions/visibilities.py]: Done

File: generic_permissions/visibilities.py (hunk 7 above)
Comment (Yelinz) [reply generic_permissions/visibilities.py]: Done

File: generic_permissions/visibilities.py (hunk 8 above)
Comment (Yelinz) [reply generic_permissions/visibilities.py]: Done

**Review by anehx — APPROVED** [review]
